new_audit: third party facades

adamraine opened the pull request:
Audit to identify third party resources which can be lazy loaded with a facade alternative.

[Doc](https://docs.google.com/document/d/1oqnFTmhpccx8ca-H_mFq40jtR4oap4Rhmf1kBPX1jeg/edit#heading=h.auhw4oaqjyyl)

Latest design (as of nov 12 2020):
![image](https://user-images.githubusercontent.com/39191/99002799-e73a2d80-24f1-11eb-96f9-86573c835c0d.png)



test page: https://www.smartling.com/

Notes on design:
* ~Having the total transfer size / total blocking time in the same row as the facade alternative is confusing. The user may get the impression that those are the individual transfer size / blocking time for the facade not the total. It may be better to not display any totals in that row to avoid this problem.~ _Resolved_
* ~It is unclear what entity each link is a part of which could be useful information.  I think we could add an extra column to hold the entity name or product name to clarify the purpose of each resource.~ _Resolved_
* ~What is the "Show 3rd party resources" checkbox for.  Is there a way to remove it?~ _Resolved_

Other notes:
* ~This PR updates ThirdPartyWeb to 0.12.2~ https://github.com/GoogleChrome/lighthouse/pull/11469
* ~The `LazyThirdParty.getSummaries()` function has a lot of shared code with the changes in https://github.com/GoogleChrome/lighthouse/pull/11219#pullrequestreview-469830821.  When that PR lands I should be able to use `ThirdPartySummary.getSummaries()` to eliminate some of the code duplication.~ _PR merged_
* ~Moving `columnBlockingTime` from `ThirdPartySummary.UIStrings` to `i18n.UIStrings` forced changes to \~50 locale files. Is there a way to avoid this?~ _Resolved_

connorjclark commented:
> Having the total transfer size / total blocking time in the same row as the facade alternative is confusing. The user may get the impression that those are the individual transfer size / blocking time for the facade not the total. It may be better to not display any totals in that row to avoid this problem.

You can accomplish this with `{key: null}` in the heading.

> It is unclear what entity each link is a part of which could be useful information. I think we could add an extra column to hold the entity name or product name to clarify the purpose of each resource.

For the subitems, yes? +1

> What is the "Show 3rd party resources" checkbox for. Is there a way to remove it?

https://github.com/GoogleChrome/lighthouse/blob/7d685217df0083bc69ffd2e8360ea7a75dd8dbb9/lighthouse-core/report/html/renderer/report-ui-features.js#L218

_____

This could be made into an opportunity by extending the ByteEfficiencyAudit, and returning a `wastedBytesByUrl` that transforms the 3p url sizes to the size of the facades. You might choose to do that in a second pass, but if so I recommend landing this in experimental in the interim.

I'm unclear on the usage of `subitems` here. It seems to map a facade to a "product type", but my assumption was that facades would be made specifically for a product.

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,183 @@
+/**
+ * @license Copyright 2019 The Lighthouse Authors. All Rights Reserved.
```
```suggestion
 * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
```

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,183 @@
+const MainThreadTasks = require('../computed/main-thread-tasks.js');
+
+const UIStrings = {
+  /** Title of a diagnostic audit that provides details about the third-party code on a web page that can be lazy loaded. This descriptive title is shown to users when no resources have lazy loading alternatives available. */
```
describe what lazy loading is in the translator string ... should probably tack that at the end of all these translator comments.

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,183 @@
+        text: str_(i18n.UIStrings.columnBlockingTime), subItemsHeading: {key: 'blockingTime'}},
+    ];
+
+    if (!results.length) {
```
nit: move to before headings

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,183 @@
+  static getSummaries(networkRecords, mainThreadTasks, cpuMultiplier, mainEntity) {
+    /** @type {Map<ThirdPartyFacade, FacadeSummary>} */
+    const summaries = new Map();
+    const defaultStat = {mainThreadTime: 0, blockingTime: 0, transferSize: 0};
```
stat -> summary, since these are `FacadeSummary`s and the resultant map is `summaries` ?

connorjclark replied:
(oh, you'll end up reusing this when that other pr lands, nvm)

adamraine commented:
@connorjclark The current iteration lists individual resources which are used by a certain product, some products have more than one.  Additionally, some facades can be made for multiple products, React Live Chat loader can be used with 4 different live chat clients for example.  However, it is unlikely a page will ever use more than one live chat client and that is the only facade which applies to multiple products for now.

Do you think listing products rather or individual resources for each facade would be better?  If listing products is better then we can eliminate the subrows for now.

adamraine commented:
Update:
<img width="892" alt="Screen Shot 2020-08-20 at 3 44 10 PM" src="https://user-images.githubusercontent.com/6752989/90817566-0151de80-e2fc-11ea-8d26-1cf49565f36a.png">

Temporary "Learn more" link: https://web.dev/efficiently-load-third-party-javascript/?utm_source=lighthouse&utm_medium=cli

patrickhulce commented:
Hm, @adamraine we might want to tweak how we surface the savings to avoid underreporting since the product request identification is only for the widget request itself and not everything it brings in.

Any ideas for ways we can restructure to surface all the work of the entity once a product is detected?

adamraine commented:
I think turning this into an opportunity should be done on a second pass because Third Party Web might need to hold the sizes for facade alternatives. With this audit in experimental I think it's ready for review:
<img width="892" alt="Screen Shot 2020-08-24 at 2 05 02 PM" src="https://user-images.githubusercontent.com/6752989/91079987-d07eeb80-e612-11ea-84a7-bc5e2675a1ea.png">

benschwarz commented:
@adamraine, I don't have anything useful to add; wanted to say: 

- Thanks for the shout of React Live Chat Loader
- Awesome to see the term "Facade Alternative", I wish I'd come up with that myself 😁

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+    const tasks = await MainThreadTasks.request(trace, context);
+    const multiplier = settings.throttlingMethod === 'simulate' ?
+      settings.throttling.cpuSlowdownMultiplier : 1;
+    const {byURL} = ThirdPartySummary.getSummaries(networkRecords, tasks, multiplier);
```
`byURL` isn't a clear variable name. just set to `thirdPartySummaries` and use `thirdPartySummaries.byURL`?

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+    for (const [url, urlSummary] of byURL) {
+      const entity = thirdPartyWeb.getEntity(url);
+      const product = thirdPartyWeb.getProduct(url);
+      if (!entity || mainEntity && entity.name === mainEntity.name) continue;
```
I think there's a helper function for this already, `isThirdParty`, which me and @patrickhulce planned on putting in a common place I think

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+      /** @type {Map<string, ProductSummary>} */
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
+      if (product && product.facades && product.facades.length) {
+        // Record new url if product has a facade
```
```suggestion
        // Record new url if product has a facade.
```

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
```
let's also define some key terms needed for understanding the code: facade, product, entity, etc...

adamraine replied:
Is there a formal way to define terms in jsdoc?

connorjclark replied:
nope

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+
+    const summary = {wastedBytes: 0, wastedMs: 0};
+
+    /** @type LH.Audit.Details.TableItem[] */
```
braces ;)

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+      const product = productSummary.product;
+      if (!product.facades || !product.facades.length) continue;
+
+      // The first facade should always be the best one
```
```suggestion
      // The first facade should always be the best one.
```

this is guaranteed by third-party-web?

adamraine replied:
Yes

connorjclark commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,189 @@
+        transferSize += urlStats.transferSize;
+        blockingTime += urlStats.blockingTime;
+      }
+      items.sort((firstEl, secondEl) => {
+        return secondEl.transferSize - firstEl.transferSize;
+      });
```
```suggestion
      items.sort((a, b) => b.transferSize - a.transferSize);
```

imo naming should be avoiding in sort fns.

adamraine commented:
@connorjclark I created a smoke test but there is some variance to the transfer size of each request, which also means the order of the results is not consistent.

Would it make sense to test the smoke test results for the number of requests found without checking the content of each request?

Another option would be to make a unit test with an example trace.

brendankenny commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,198 @@
+      } else {
+        // If the url does not have a facade but one or more products on its entity do,
+        // we still want to record this url because it was probably fetched by a product with a facade.
+        for (const productSummary of productSummaries.values()) {
```
is this assuming an ordering of `byURL` that if this url was part of a product that a summary for it would already exist in `productSummaries`?

adamraine replied:
Good point, this should definitely be changed

adamraine replied:
Interestingly, I noticed the results were the same after changing this (i.e. we weren't missing any URLs that happened to appear before the product URL in `byURL`.

I didn't know that `Map()` remembers to order of insertion. The product URL will always appear first in the network records so the entries will always be in an order where the `productSummary` is created first.

brendankenny replied:
> I didn't know that `Map()` remembers to order of insertion. The product URL will always appear first in the network records so the entries will always be in an order where the `productSummary` is created first.

Ah, got it, but that does assume network records are sorted correctly. That's probably usually true, but they are created from DevTools events and I don't think we ever explicitly sort them, so you never know.

This could sort `byURL` before this loop or this could take two passes (generate productSummaries, then associate URL with them), but that might be complete overkill. cc @patrickhulce to provide a check on my possibly worrying about nothing :)

adamraine replied:
I have a solution that does two passes from when I was testing this, I don't think it will hurt to use it either way.

brendankenny reviewed (commented):
quick initial review, but I'll dig in from here:

- As a user, do I care about the individual parts of the youtube embed and the cost of each, or should we just show the iframe URL and totals for all its sub-resources?
- I assume for OOPIFs, `mainThreadTime`/`blockingTime` will always be zero (at least for iframe embeds). Are we underselling the performance impact in those cases? (i.e. why should I care about the audit if the main thread impact is apparently zero?) There's also the time it takes to download the assets (possibly blocking other downloads) and time for it to then paint something to screen, making it look like the page isn't ready in the meantime, etc

> With this audit in experimental

is this supposed to be in experimental or was that changed?

adamraine commented:
> As a user, do I care about the individual parts of the youtube embed

One reason to list individual parts is to make sure the user knows the cost of including a resource also includes sub-resources. Listing the total cost next to the iframe URL may be a bit misleading in this way but could be explained in labeling and the audit description.

> I assume for OOPIFs, `mainThreadTime`/`blockingTime` will always be zero (at least for iframe embeds). Are we underselling the performance impact in those cases? (i.e. why should I care about the audit if the main thread impact is apparently zero?) There's also the time it takes to download the assets (possibly blocking other downloads) and time for it to then paint something to screen, making it look like the page isn't ready in the meantime, etc

I think we are certainly underselling in the current state. We can turn this into an opportunity and compute savings similar to `render-blocking-resource` and `byte-efficiency-audit`. I was originally going to do this in a second pass because of Third Party Web changes, but those changes may not be necessary anymore. (https://github.com/GoogleChrome/lighthouse/issues/11344).

> > With this audit in experimental
>
> is this supposed to be in experimental or was that changed?

This was changed, it's supposed to be in default now

connorjclark commented:
> @connorjclark I created a smoke test but there is some variance to the transfer size of each request, which also means the order of the results is not consistent.
> 
> Would it make sense to test the smoke test results for the number of requests found without checking the content of each request?
> 
> Another option would be to make a unit test with an example trace. This PR doesn't make any changes to gatherers so perhaps an e2e smoke test is not necessary.

We've run into this before, and made a similar compromise of just checking lengths ... I'm starting to think we should have something in smokehouse to normalize the order of arrays–just spitballing:

```
{
  arrayValue: [
     {_smokehouse: {sortBy: 'time'}},
     {actualData: 'blah, time: 12},
     {actualData: 'blah, time: 14},
    {actualData: 'blah, time: 16},
  ],
}
```

I would prefer a smoketest over a trace test, but given the work needed to support the above, trace is okay with me.

brendankenny commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,201 @@
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    const settings = context.settings || {};
```
there's always a `settings`, no need for a fallback

brendankenny commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,201 @@
+    const results = [];
+    for (const productSummary of productSummaries) {
+      const product = productSummary.product;
+      if (!product.facades || !product.facades.length) continue;
```
shouldn't `getProductSummaries()` have not created summaries without facades?

brendankenny commented on lighthouse-core/test/audits/lazy-third-party-test.js:
```diff
@@ -0,0 +1,131 @@
+
+  it('only reports resources which can be lazy loaded', async () => {
+    const artifacts = {
+      devtoolsLogs: {defaultPass: pwaDevtoolsLog},
```
worth adding a comment why this devtoolsLog/trace might trigger the audit otherwise? I assume it has third party requests, but they don't have facades?

brendankenny commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,201 @@
+   * @param {ThirdPartyEntity | undefined} mainEntity
+   * @return {ProductSummary[]}
+   */
+  static getProductSummaries(byURL, mainEntity) {
```
nit: maybe something like `getFacadableProductSummaries`?

brendankenny commented on lighthouse-core/test/audits/lazy-third-party-test.js:
```diff
@@ -0,0 +1,131 @@
+        ]},
+      },
+    ]);
+  });
```
I know this is getting lengthy (maybe there's a way to share generated test data), but it would be great to get a few more tests exercising things like items before the cutoff aren't included in the results, network requests for a product interleaved with requests from other entities, more than one of the same product, more than one product in the same load. Not all of these are necessarily important, and some can be combined, but good to get coverage, especially so future refactors can trust the tests and not worry about e.g. url ordering all over again :)

brendankenny reviewed (commented):
looking at sites with multiple youtube embeds, it seems like the desired behavior is to sum all those embeds in a single entry, right? (just checking)

One weird site: https://www.polygon.com/2020/9/9/21428944/xbox-series-x-s-launch-lineup-games

It does show the multiple embeds rolled up into one entry, but the vast majority of the embeds' contents aren't being counted (e.g. the giant youtube `base.js` is indeed in the `network-requests` audit in the LHR, so it happened, but it isn't being caught in the lazy-third-party-results):

<img width="898" alt="lazy-third-party audit multiple embed results" src="https://user-images.githubusercontent.com/316891/92975879-e3851e80-f44e-11ea-8b54-ef8f09eed983.png">

I thought maybe it was the multiple embeds, but the same thing happens on one of their pages with a single embed (like [this one](https://www.polygon.com/2020/9/10/21431181/ubisoft-new-game-riders-republic-extreme-sports-ps4-ps5-xbox-one-xbox-series-x-pc)), though this time it does catch the ytimg thumbnail:

<img width="894" alt="lazy-third-party audit single embed results" src="https://user-images.githubusercontent.com/316891/92976030-38c13000-f44f-11ea-8ebc-f32043cd14fe.png">

adamraine commented:
I defined `cutoff` as the `endTime` of the original product request, in the case of YouTube embed its a request to url `www.youtube.com/embed/...`. Any request which has a `startTime` after `cutoff` is plausibly fetched by that product. On the pages you gave, some of the network records for the YouTube resources have a `startTime` **before** `cutoff` and an `endTime` which is **after** `cutoff`.

If I adjust the audit to include requests with `endTime` after `cutoff` instead of `startTime`, everything appears to work as expected. These are the results for the first link with multiple embeds:

<img width="877" alt="Screen Shot 2020-09-11 at 7 01 06 PM" src="https://user-images.githubusercontent.com/6752989/92979574-2c45d300-f461-11ea-9b9a-43c00634e5b5.png">

Similar case with the second link and single embed:

<img width="877" alt="Screen Shot 2020-09-11 at 7 09 05 PM" src="https://user-images.githubusercontent.com/6752989/92979951-4633e580-f462-11ea-9766-c62513081fce.png">

Interestingly, the network requests in the trace appear to be showing the correct behavior so I will need to investigate the difference in timings between the trace and devtools log.

<img width="802" alt="Screen Shot 2020-09-11 at 6 35 58 PM" src="https://user-images.githubusercontent.com/6752989/92978424-ae33fd00-f45d-11ea-9a85-15677ff8defa.png">

patrickhulce commented:
> I defined cutoff as the endTime of the original product request, in the case of YouTube embed its a request to url

You might want to use `receiveHeadersEnd` as the cutoff instead of `endTime`. For a document, Chrome's lookahead preparser will start many resource requests before the entire HTML document has finished loading.

adamraine commented:
Good point. Additionally, it looks like multiple requests to the same URL was causing the issue here. The end time of the URL was set as the end time of the last request, not the first.

adamraine commented:
@patrickhulce `request.timing.receiveHeadersEnd` is on a different time scale than `request.endTime`. As a result `request.timing.receiveHeadersEnd` cannot be directly compared to `request.startTime` when determining the order of requests for this audit.

Does it make sense to instead compare `request.timing.receiveHeadersEnd` to something like `request.timing.dnsStart` or `request.timing.connectStart`?

patrickhulce commented:
@adamraine it's been a while since I've worked with them, but IIRC the values inside timing are actually ms *since* the `startTime`

adamraine commented on lighthouse-core/audits/lazy-third-party.js:
```diff
@@ -0,0 +1,201 @@
+   */
+  static get meta() {
+    return {
+      id: 'lazy-third-party',
```
I'm on board with this idea. If we rename to something like `third-party-facades` then I think we should change the comments and UI so they only mention lazy loading in relation to facade alternatives.

For example, the title *"Some third-party resources can be lazy loaded"* should be changed to *"Some third-party resources have a facade alternative"* or *"Some third-party resources can be lazy loaded with a facade"*.

How do you feel about this?

brendankenny replied:
> I'm on board with this idea. If we rename to something like `third-party-facades` then I think we should change the comments and UI so they only mention lazy loading in relation to facade alternatives.

Sounds good to me. Looking at the original audit planning doc, it seems like there's a pretty natural split between lazy loading via facade and lazy loading via async or after LCP or whatever.

connorjclark commented:
@adamraine and I think it would be good to not use specific facade products ie. React Live Chat Loader. Instead, we just say the genre of facades–Chat/Customer Support/etc... and then link to web.dev with all the subtly and every possible facade alternatives there listed.

Allows us to update outside of release cycles, and not have to come up with a UI for showing multiple, and not have to pick "just one" or what we think might be the best (no risk of favoritism in the report).

patrickhulce commented:
Along the lines of what @connorjclark is saying, I think it might be a good idea to bring this audit to the group and run it through whatever process we develop as a result of the large libraries audit before we land in the default config. 

I think many aspects of the facade identification here share concerns with our opinionated advice guide and we'll want to be careful rolling this one out (and probably update a fair bit of the facade data in third-party-web, if we still end up using it).

adamraine commented:
@connorjclark when you say "genre of facade", is that different from the existing categories in third party web (video, customer success, etc.)?

connorjclark commented:
> @connorjclark when you say "genre of facade", is that different from the existing categories in third party web (video, customer success, etc.)?

i just made it up, but yea that'd be the same thing

adamraine commented:
@connorjclark I've put some more thought into this updated design.

If the existing facades can be listed in web.dev behind the "Learn more" link does it really help us if the category also links to a specific part of that page. Were you thinking the third party categories would have different pages?

Additionally, the "Facade Alternative" column with just the repo link made sense because that was essential information, I think it's harder to justify keeping a weird single-row column like that for something like the category.

If having the category is important, I think we should put it in the product column without linking to anything or have the product name column link to web.dev. Could look like this:
```
Product                            Transfer size           Blocking time

YouTube Embed (Video)               600Kb                   0ms
   youtube.com/embed/...            40Kb                    0ms
```

connorjclark commented:
> If the existing facades can be listed in web.dev behind the "Learn more" link does it really help us if the category also links to a specific part of that page.

It may help if there's more than a handful of products. And having a link inline (in the table results) is good too, some users might have a UX-blindness to a "Learn more" link.

> Were you thinking the third party categories would have different pages?

No, one page.

> If having the category is important, I think we should put it in the product column without linking to anything or have the product name column link to web.dev. Could look like this:
> 
> ```
> Product                            Transfer size           Blocking time
> 
> YouTube Embed (Video)               600Kb                   0ms
>    youtube.com/embed/...            40Kb                    0ms
> ```

This is how I imagined it (with "Youtube Embed...." being an anchor link to web.dev where we list all the ways to lazy load this product)

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,227 @@
+  failureTitle: 'Some third-party resources can be lazy loaded with a facade alternative',
+  /** Description of a Lighthouse audit that identifies the third party code on the page that can be lazy loaded with a facade alternative. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. Lazy loading means loading resources is deferred until they are needed. */
+  description: 'Some third-party resources can be fetched after the page loads. ' +
+    'These third-party resources are used by embedded elements which can be replaced by a facade ' +
```
marking this string as needs-editing (don't have a good suggestion atm tho)

adamraine replied:
Some more ideas:

- Some third party embeds can be replaced by a facade until they are required.
- Some third party embeds can be lazy loaded. You can replace them with a facade until they are required.

connorjclark replied:
I like the second option, thanks.

adamraine commented:
Current look:
<img width="895" alt="Screen Shot 2020-09-21 at 6 29 39 PM" src="https://user-images.githubusercontent.com/6752989/93828002-70db2680-fc38-11ea-9906-8ceeec98dfb0.png">

connorjclark commented on lighthouse-core/test/audits/third-party-facades-test.js:
```diff
@@ -0,0 +1,514 @@
+      ]
+    );
+  });
+});
```
can you add some tests where blockingTime is not 0?

adamraine replied:
Possibly not, but I will investigate. Iframe widgets like YouTube and Vimeo will always have a blocking time of 0. The live chat widgets like intercom which do have non-zero blocking time may not be free to use.

connorjclark commented:
Adam pointed out that this is blocked on web.dev documentation, which he is about to start.

adamraine commented:
Linking the web.dev draft [here](https://docs.google.com/document/d/1WCDboioqEuoBGLKQ0UDe9-ptg41lEPk9-93wzTWrZ6M/edit?usp=sharing).

I figured we should discuss it since it's going to be an integral part of this audit.

connorjclark commented:
1) Could we omit "0" from the blocking time column, to reduce the noise in the table?
2) Should we aggregate assets <1KB into a single row? (Aside: I'd love for the table/report to be able to do this itself, and allow for expanding an "Other" row. currently, doing it manually in the audit is what we do)

adamraine commented:
> Could we omit "0" from the blocking time column, to reduce the noise in the table?

What do you mean? Put nothing in the column if the blocking time is 0? Hide the column if all blocking times are 0?

> Should we aggregate assets <1KB into a single row? 

Third party summary audit does something similar but it only shows the top 5 resources and combines everything else into an "Other" row. Does it make more sense to do top 5 resources or all resources >=1KB?

connorjclark commented:
> Put nothing in the column if the blocking time is 0

This. I believe a value of "null" would do the trick. @patrickhulce does this seem bad or ok?

> Does it make more sense to do top 5 resources or all resources >=1KB?

Either/or, I'll leave it up to you.

adamraine commented:
@connorjclark How's this? Should we also hide the blocking time column if there is nothing in it?

<img width="899" alt="Screen Shot 2020-09-29 at 5 16 51 PM" src="https://user-images.githubusercontent.com/6752989/94617451-963ee480-0277-11eb-9277-a6f42d30966e.png">

connorjclark commented:
I think it looks better. The only issue one might have is that data representation. Are ppl okay with nulls in your numbers? :) 

> Should we also hide the blocking time column if there is nothing in it?

Let's no go that far.

adamraine commented:
I had to use `undefined` instead of `null` btw

brendankenny commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,241 @@
+    const productSummaries
+      = ThirdPartyFacades.getFacadableProductSummaries(thirdPartySummaries.byURL, mainEntity);
+
+    const summary = {wastedBytes: 0, wastedMs: 0};
```
fwiw I don't think table `summary` is used by anything

brendankenny commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,241 @@
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity || thirdPartyWeb.isFirstParty(url, mainEntity)) continue;
+
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
```
I believe you need a `/** @type {Map<string, FacadableProductSummary>} */` here or it'll be `Map<any, any>`

brendankenny replied:
move inside the `product` conditional since there won't be any changes to `productSummaries` if there's no product to add?

brendankenny commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,241 @@
+    /** @type {Map<string, Map<string, FacadableProductSummary>>} */
+    const entitySummaries = new Map();
+
+    for (const [url, urlSummary] of byURL) {
```
comments for why there are two loops would be good. "First find the urls associated with facadable products something something", then below, "Find urls not associated with a product but likely part of the product's resources" or something

brendankenny commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,241 @@
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity || thirdPartyWeb.isFirstParty(url, mainEntity)) continue;
+
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
```
if there's no existing `productSummaries` this can just `continue`, right? (and no need to re `.set()` it below)

brendankenny commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,241 @@
+        /** @type {FacadableProductSummary} */
+        const productSummary = productSummaries.get(product.name) || {
+          product,
+          cutoffTime: Infinity,
```
should this be named something with `start` in it? Since we are only going to associate things with it _after_ this time, but `cutoff` sounds like the end of something (to me, at least :)

adamraine replied:
I thought that might get mixed up with `startTime` since `cutoff` is the `endTime`-ish of some request, maybe `beginning` works better?

patrickhulce commented on lighthouse-core/audits/third-party-summary.js:
```diff
@@ -85,18 +99,27 @@ class ThirdPartySummary extends Audit {
       urlSummary.transferSize += request.transferSize;
+      urlSummary.firstStartTime = Math.min(urlSummary.firstStartTime, request.startTime);
+      if (request.timing) {
+        // Lookahead preparser may start requests before entire HTML is loaded
```
I'm not really sure how this comment applies here either. Feels like we should just attach the earliest network request object for the URL at this point rather than store these other-audit-specific summary attributes?

adamraine replied:
In theory could we have two requests to the same URL but the request which starts second but ends first? If we store a single request we cannot capture this situation.

patrickhulce replied:
> In theory could we have two requests to the same URL but the request which starts second but ends first?

What happens in the audit if that situation occurred and we used the earlier one?

adamraine replied:
```
P1 =====================================
P2       ========     R1 ========
```

It could mean a slightly different cutoff time for resources fetched by the product. The request which ends first (`P2`) might be used to start fetching secondary resources (`R1`) before the other request (`P1`) ends. If we use the request which ends second (`P1`), all these secondary resources (`R1`) would not be counted.

This scenario could be avoided if we store based on first end time, but that leads to a different problem:
```
P1 =================
         R1 ===========================
                              R2 ============
```
These two requests to the same resource (`R1`, `R2`) would be counted even though they shouldn't.

Both situations seem pretty unlikely (maybe impossible?) but to avoid both we can't take the start time and end time from the same request.

patrickhulce replied:
Ah thanks for the explanation :) I see my confusion now. I thought we should just be showing all the assets from the entity under the product anyway, so I was never really concerned with this problem. At least as far as I could tell, the product request in third-party-web identified was never *guaranteed* to be the *first* request related to that product anyway, it was just *definitely* a request for that product. 

Was this your understanding too or was there more work than I realized that the product labelled request is always guaranteed to the be first request related to this product?

(A key example that comes to mind is the Google Analytics loader which tells you to load it through GTM even if that's the only thing you're loading. In that case, none of our proposed handling methods will ever correctly attribute that Google Analytics is the only reason you're requesting this third-party.)

>

adamraine replied:
The product regex in TPW is used to match the FIRST request for a specific product, but any request fetched after this initial product request that belongs to the same entity is also included in the report. The URL for this initial product request is the URL the developer sees when they add one of these third party products (e.g. the source url in the iframe for a YouTube embed).

patrickhulce replied:
> The product regex in TPW is used to match the FIRST request for a specific product, but any request fetched after this initial product request that belongs to the same entity is also included in the report.

I think we're talking about different things right now :) Taking a step back...

I'm asking as maintainer of third-party-web if I'm supposed to be upholding a contract that all regexes that appear under a `product` identifier must only identify the first request that is ever issued for that product, which wasn't my understanding when we added the product identification there. I thought it was just marking any URL that definitely identified usage of that product on the page. 

If my understanding *was* correct, then I'm hesitant to use that signal here as the end-all-be-all of when the product-related activity started even if it works out in the YouTube case (because also just using the entity works out for the most part too 😉 ). Does that make sense?

If my understanding was wrong, we might want to consider moving this rather Lighthouse-specific "first product request" identification logic over here into Lighthouse rather than in the generic TPW dataset and sorry for leading you astray!

adamraine replied:
I think I see what you're saying now. Your understanding is different from mine, the regexes I put in TPW were meant to be the first request as well as the identifying request.

Moving the first request logic to LH is fine with me. I would still argue that the identifying resource can be used as the starting point for resources that will always be deferrable. However, if there is only one product under its entity then some other entity resources could be deferrable.

What do you think about having exhaustive product lists under entities? This would allow us to recommend deferring the entire entity if there is a single deferrable product vs deferring a single product under that entity.

brendankenny replied:
@adamraine @patrickhulce it's hard to tell where this conversation ended up :)

patrickhulce replied:
> What do you think about having exhaustive product lists under entities? 

If I'm understanding your proposal correctly of ensuring all URLs of each third-party are assigned to a product, seems like an impossible amount of work to commit to. Domain ownership is easier to track and slower moving but tracking slight internal URL changes and dependencies would need an army of volunteers to maintain third-party-web. Finding *one* request that definitely detects a product was already a big step on its own 😅

> it's hard to tell where this conversation ended up

I'm not quite sure either, tbh. I think we're mostly splitting hairs about something that won't have *too* much impact on the results. 


Summary of my position:

1. I'd like to keep the identification logic in third-party-web inline with that project's mission, to identify domains/URLs that definitely belong to a particular entity or product and not let Lighthouse-specific sequencing-based concerns leak into contributions that can or can't be accepted there.
2. I want to surface the *complete* impact of a third-party product on a page here in Lighthouse.
3. I believe just surfacing all of an entity's resources once a product is detected is the simplest way to accomplish two and has minimal risk of overstating the product's impact. 

I think @adamraine and I are on the same page for 1 and 2, and maybe we just disagree about 3?

adamraine replied:
That pretty much sums it up. Patrick is right that splitting resources based on timing won't have too much impact, but it seems like a pure positive aside from the code organization problem that was the original issue in this thread. That being said, I have not found a site where the impact would be significantly different so I think I will get rid of the timings.

adamraine commented on lighthouse-core/lib/third-party-web.js:
```diff
@@ -23,6 +24,20 @@ function getEntity(url) {
+  try {
+    return thirdPartyWeb.getProduct(url);
+  } catch (_) {
+    return undefined;
```
Were we talking about getting rid of the try block entirely? I can't find a situation where `getProduct` or `getEntity` will throw.

brendankenny replied:
> Were we talking about getting rid of the try block entirely? I can't find a situation where `getProduct` or `getEntity` will throw.

Based on the comment it sounds like they threw on invalid URLs, but maybe that changed in more recent versions of `third-party-web`?

adamraine replied:
@patrickhulce do you know what would cause `getProduct` or `getEntity` to throw?

patrickhulce replied:
originally it would throw if the provided URL was completely invalid, I [since updated it to return `null` in that case](https://github.com/patrickhulce/third-party-web/commit/fa6e9888012a29dab76c4c3f5b79619643675039#diff-cd3025771cb9802302b141b004bef6c1). conceivably there could be another unexpected case in third-party-web that throws we're not aware of, but that would be considered a bug and not really worth of a try/catch :)

adamraine replied:
I removed both try blocks.

adamraine commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,241 @@
+        productWithCategory = str_(categoryTemplate, {productName: product.name});
+      } else {
+        // Just display product name if no category is found
+        productWithCategory = product.name;
```
No products hit this fallback.

adamraine commented on lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:
```diff
@@ -354,4 +354,31 @@ module.exports = [
     },
   },
+  {
+    lhr: {
```
How do you feel about a new category like `perf-serial` and move every test which needs to be run serially in that group?

brendankenny replied:
> How do you feel about a new category like `perf-serial` and move every test which needs to be run serially in that group?

to me it makes more sense to keep the metrics tests as `perf` (since they're the core of the perf category) and split these perf opportunities/diagnostics into their own name (and `yarn smoke perf-serial` is a lot more awkward to type :), but I don't feel super strongly about it.

adamraine replied:
Ok I think I will open an issue for this and tackle in a separate PR.

adamraine commented on lighthouse-core/audits/third-party-summary.js:
```diff
@@ -41,21 +37,31 @@ const PASS_THRESHOLD_IN_MS = 250;
+ * @typedef Summary
+ * @property {number} mainThreadTime
+ * @property {number} transferSize
+ * @property {number} blockingTime
```
There are a lot of style changes in this file with no accompanying functional changes anymore. Is it worth keeping the style changes anyway?

connorjclark replied:
these changes are good, lets keep

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,214 @@
+  'social': UIStrings.categorySocial,
+};
+
+/** @typedef {import("third-party-web").IEntity} ThirdPartyEntity */
```
nit: move typedefs to just below fileoverview

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,214 @@
+   * @param {ThirdPartyEntity | undefined} mainEntity
+   * @return {FacadableProduct[]}
+   */
+  static getFacadableProducts(byURL, mainEntity) {
```
getProductsWithFacades ?

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,214 @@
+      settings.throttling.cpuSlowdownMultiplier : 1;
+    const summaries = ThirdPartySummary.getSummaries(networkRecords, tasks, multiplier);
+    const facadableProducts
+      = ThirdPartyFacades.getFacadableProducts(summaries.byURL, mainEntity);
```
nit: `=` should be just before newline

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  /**
+   * @param {ThirdPartySummary.URLSummary[]} items
```
note the precondition that items is sorted by transferSize; or just do the sort inside this function?

adamraine replied:
Yeah, doing sort inside the function sounds better.

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,214 @@
+    const headings = [
+      /* eslint-disable max-len */
+      {key: 'product', itemType: 'text', subItemsHeading: {key: 'url', itemType: 'url'}, text: str_(UIStrings.columnProduct)},
+      {key: 'transferSize', granularity: 1, itemType: 'bytes', subItemsHeading: {key: 'transferSize'}, text: str_(i18n.UIStrings.columnTransferSize)},
```
super nit: use order `key, itemType, subItemsHeading, granularity, text`. order is by usefulness

adamraine commented:
@connorjclark @brendankenny Chrome tech writers are working on publishing the documentation. Can you take another look through?

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,209 @@
+const ThirdPartySummary = require('./third-party-summary.js');
+
+const UIStrings = {
+  /** Title of a diagnostic audit that provides details about the third-party code on a web page that can be lazy loaded with a facade alternative. This descriptive title is shown to users when no resources have facade alternatives available. Lazy loading means loading resources is deferred until they are needed. */
```
I think "third party" will need to be explained in the translation comment.

connorjclark replied:
starting to wonder if we should add a glossary to these comments

```
/**
 * @glossary
 * @word {Facade} Something blah blah blah
 * @word {Product} Something blah blah blah
 * @word {Third-Party} Something blah blah blah
 */
```

And then concat all of these definitions to every translator comment in this file when we collect strings.

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,209 @@
+  /**
+   * @param {ThirdPartySummary.URLSummary[]} items
+   */
+  static condenseItems(items) {
```
jsdoc comment saying what this does would be useful

connorjclark commented on lighthouse-core/lib/third-party-web.js:
```diff
@@ -8,19 +8,22 @@
  * @return {ThirdPartyEntity|undefined}
  */
 function getEntity(url) {
-  try {
```
Does this not throw an error anymore? cc @patrickhulce

adamraine replied:
I don't think it throws anymore https://github.com/GoogleChrome/lighthouse/pull/11290#discussion_r498364311

patrickhulce replied:
yep at least not in ways we expect :)

patrickhulce commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -108,6 +110,8 @@ const UIStrings = {
   /** Label for a row in a data table; entries will be the total number and byte size of all third-party resources loaded by a web page. 'Third-party resources are items loaded from URLs that aren't controlled by the owner of the web page. */
   thirdPartyResourceType: 'Third-party',
+  /** Label used to identify a value in a table where many individual values are aggregated to a single value, for brevity. "Other resources" could also be read as "the rest of the resources". Resource refers to network resources requested by the browser. */
+  otherValue: 'Other resources',
```
Now that it's here in a generic place, WDYT about `subItemOther` or something more descriptive for how it's used?

connorjclark replied:
why specify subItem? I don't think the text would change for a row.

patrickhulce replied:
OK, `rowOther` then?

I don't think we should ever create a non-subitem row that says this though, so now that we're on the subject, I actually like `subItemOther` even more :)

adamraine replied:
I'm going with `otherResourcesLabel` since the text is specific to "resources"

patrickhulce commented on lighthouse-core/test/audits/third-party-facades-test.js:
```diff
@@ -0,0 +1,450 @@
+const videoEmbedsDevtolsLog = require('../fixtures/traces/video-embeds-m84.devtools.log.json');
+const noThirdPartyTrace = require('../fixtures/traces/no-tracingstarted-m74.json');
+
+function resourceEntry(startTime, headerEndTime, endTime, transferSize, url) {
```
FWIW, I've come to regret *almost* every test helper API I've made where there are no names whatsoever 

I'm not asking to change it, but from someone who has written many of these in LH, just a nudge of personal experience to consider switching sides and consider using sensible defaults but keep the object names instead 😁

e.g. derive headerEndTime as startTime +1 for all the tests that don't care, derive endTime as startTime +2, pass in `{startTime: 100, url: '...'}`

<img src="https://user-images.githubusercontent.com/2301202/98748716-d36eba00-237f-11eb-9f39-e8fed069da09.png" width="200">

adamraine replied:
The biggest reason I made the test helpers was to make the request timing easy to understand. Now that request timing doesn't matter anymore, I think this change is worth making.

patrickhulce commented on lighthouse-core/test/audits/third-party-facades-test.js:
```diff
@@ -0,0 +1,450 @@
+ */
+'use strict';
+
+const ThirdPartyFacades = require('../../audits/third-party-facades.js');
```
awesome extensive tests! 👍

patrickhulce commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,209 @@
+   */
+  static condenseItems(items) {
+    items.sort((a, b) => b.transferSize - a.transferSize);
+    const splitIndex = items.findIndex((item) => item.transferSize < 1000);
```
feels like we could roll up a little more, this index or a max of 6 or something similar to the other third party audit?

patrickhulce commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,209 @@
+    if (splitIndex === -1) return;
+
+    const remainder = items.splice(splitIndex);
+    const finalItem = remainder.reduce((result, item) => {
```
WDYT about not showing this if the remainder is still under 1000?

patrickhulce reviewed (approved):
LGTM, nicely done! 🎉

adamraine commented on lighthouse-cli/test/fixtures/perf/third-party.html:
```diff
@@ -0,0 +1,14 @@
+<html>
+<body>
+
+<div>AAAAAAAAAAAAAAAAAAA</div>
```
Guilty as charged

brendankenny reviewed (approved):
LGTM2! Nice work.

adamraine commented on lighthouse-core/lib/third-party-web.js:
```diff
@@ -8,19 +8,22 @@
+ * @param {string} url
+ * @return {ThirdPartyProduct|undefined}
+ */
+function getProduct(url) {
```
We make use of the `isFirstParty` function in `third-party-facades.js`. I would rather not reimplement that or use two different `thirdPartyWeb` modules so I prefer leaving this TBH.

patrickhulce replied:
Also non-trivial to remember to fetch the *same* no stats bundle subset to avoid duplication so I'd also prefer to have all 3p usages come through here 👍

adamraine commented:
@brendankenny @patrickhulce @connorjclark Thanks for all the comments! There were a lot of them so could ya'll take another look to make sure I properly addressed your concerns :).

patrickhulce reviewed (approved):
still LGTM!

connorjclark commented on lighthouse-core/audits/third-party-facades.js:
```diff
@@ -0,0 +1,220 @@
+  failureTitle: 'Some third-party resources can be lazy loaded with a facade',
+  /** Description of a Lighthouse audit that identifies the third-party code on the page that can be lazy loaded with a facade alternative. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. A facade is a lightweight component which looks like the desired resource. Lazy loading means resources are ...
+  description: 'Some third-party embeds can be lazy loaded. ' +
+    'Consider replacing them with a facade until they are required. [Learn more](https://web.dev/efficiently-load-third-party-javascript/).',
```
We're going to wait for docs before landing, right?

adamraine replied:
Correct

adamraine replied:
```suggestion
    'Consider replacing them with a facade until they are required. [Learn more](https://web.dev/third-party-facades/).',
```
